_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,681 @@
+is inserted at the beginning, causes a shift of all other declarations towards the end. At least now our alarm bell
+should ring, and it's time to reconsider our test case!
+
+### Avoid `Mockito.reset` for Better Unit Tests
+
+Mockito recommends in their documentation to prefer recreation of mocks over resetting them.
+
+> Smart Mockito users hardly use this feature because they know it could be a sign of poor tests.
+> Normally, you don't need to reset your mocks, just create new mocks for each test method.
+
+We better create simple and small test cases than lengthy and over-specified tests. The cause of such tests might be
+testing too much in a single unit test. But let's see an example for this situation.
+
+```java
+  @Test
+  void findAndDelete() throws ElementNotFoundException {
+    City expected = createCity();
+    Mockito.when(cityRepository.find(expected.getId()))
+      .thenReturn(Optional.of(expected));
+    City actual = cityService.find(expected.getId());
+    ReflectionAssert.assertReflectionEquals(expected,actual);
+    cityService.delete(expected);
+    Mockito.verify(cityRepository).delete(expected);
+    Mockito.reset(cityRepository);
+    Mockito.when(cityRepository.find(expected.getId()))
+      .thenReturn(Optional.empty());
+    Assertions.assertThrows(ElementNotFoundException.class,
+      () -> cityService.find(expected.getId()));
+  }
+```
+
+What does this test case?
+
+1. Tries to find a city and an assertion
+2. Deletes a city and a verify
+3. Tries to find the previously city again but expecting an exception.
+
+Such a test case's design is unfortunate. It *tests* too much and could be split in simpler and smaller units.
+
+```java
+  @BeforeEach
+  void setUp() {
+    cityRepository = Mockito.mock(CityRepository.class);
+    cityService = new CityServiceImpl(cityRepository);
+  }
+  
+  @Test
+  void find() throws ElementNotFoundException {
+    City expected = createCity();
+    Mockito.when(cityRepository.find(expected.getId())).thenReturn(Optional.of(expected));
+    City actual = cityService.find(expected.getId());
+    ReflectionAssert.assertReflectionEquals(expected,actual);
+  }
+  
+  @Test
+  void delete() throws ElementNotFoundException {
+    City expected = createCity();
+    cityService.delete(expected);
+    Mockito.verify(cityRepository).delete(expected);
+  }
+  
+  @Test
+  void findThrows () {
+    City expected = createCity();
+    Mockito.when(cityRepository.find(expected.getId())).thenReturn(Optional.empty());
+    Assertions.assertThrows(ElementNotFoundException.class,()->cityService.find(expected.getId()));
+  }
+```
+
+Now each test is simple and easy understandable. We don't have to reset the mocks anymore, since this is achieved in
+the `setUp` method. The effectively tested code is the same but a lot more meaningful than before.
+
+### Don't Mock Collections or Value Objects
```

Comment:
Can you add a code example for this section? I think this would make it easier to understand for many readers.